File: c/Dart/CLexerBase.dart
import 'dart:io';
import 'package:antlr4/antlr4.dart';

abstract class CLexerBase extends Lexer {
  CLexerBase(CharStream input) : super(runGccAndMakeStream(input));

  static CharStream runGccAndMakeStream(CharStream input) {
    var isWindows = Platform.isWindows;

    // Get options from command line args
    var args = Platform.executableArguments + Platform.script.pathSegments;

    var vsc = args.any((a) => a.toLowerCase().contains("--vsc"));
    var gcc = args.any((a) => a.toLowerCase().contains("--gcc"));
    var clang = args.any((a) => a.toLowerCase().contains("--clang"));
    var nopp = args.any((a) => a.toLowerCase().contains("--nopp"));

    if (!(vsc || gcc || clang)) {
      gcc = true;
    }

    // Extract preprocessor options (-D and -I)
    var ppOptions = extractPreprocessorOptions(args);

    // Get the source name from the CharStream
    var sourceName = input.sourceName;
    var inputText = input.getText(Interval(0, input.size - 1));

    // If source name is empty or not a .c file, we need to write to a temp file
    if (sourceName.isEmpty || !sourceName.endsWith(".c")) {
      // Create a temp file for preprocessing
      sourceName = "${Directory.systemTemp.path}/antlr4_temp_${DateTime.now().millisecondsSinceEpoch}.c";
      File(sourceName).writeAsStringSync(inputText);
    }

    var outputName = "$sourceName.p";

    if (nopp) {
      try {
        File(outputName).writeAsStringSync(inputText);
      } catch (e) {
        // Ignore
      }
      return InputStream.fromString(inputText);
    }

    if (gcc) {
      var output = "";
      try {
        var gccCommand = isWindows ? "gcc.exe" : "gcc";
        var gccArgs = ["-std=c2x", "-E", "-C"];
        gccArgs.addAll(ppOptions);
        gccArgs.add(sourceName);
        var result = Process.runSync(gccCommand, gccArgs);
        output = result.stdout as String;
      } catch (e) {
        // Failed to run gcc preprocessor
      }
      File(outputName).writeAsStringSync(output);
      return InputStream.fromString(output);
    }
    if (clang) {
      var output = "";
      try {
        var clangCommand = isWindows ? "clang.exe" : "clang";
        var clangArgs = ["-std=c2x", "-E", "-C"];
        clangArgs.addAll(ppOptions);
        clangArgs.add(sourceName);
        var result = Process.runSync(clangCommand, clangArgs);
        output = result.stdout as String;
      } catch (e) {
        // Failed to run clang preprocessor
      }
      File(outputName).writeAsStringSync(output);
      return InputStream.fromString(output);
    }

    throw Exception("No preprocessor specified.");
  }

  static List<String> extractPreprocessorOptions(List<String> args) {
    var options = <String>[];
    for (var arg in args) {
      // Match --D and --I options
      if (arg.startsWith("--D")) {
        options.add("-D${arg.substring(3)}");
      }
      else if (arg.startsWith("--I")) {
        options.add("-I${arg.substring(3)}");
      }
    }
    return options;
  }
}

File: c/Python3/CLexerBase.py
import sys
import os
import platform
import subprocess
from antlr4 import Lexer, InputStream

class CLexerBase(Lexer):
    def __init__(self, input, output=sys.stdout):
        super().__init__(CLexerBase.runGccAndMakeStream(input), output)

    @staticmethod
    def runGccAndMakeStream(input):
        is_windows = platform.system() == "Windows"

        args = sys.argv

        vsc = any("--vsc" in a.lower() for a in args)
        gcc = any("--gcc" in a.lower() for a in args)
        clang = any("--clang" in a.lower() for a in args)
        nopp = any("--nopp" in a.lower() for a in args)

        if not (vsc or gcc or clang):
            gcc = True

        ppOptions = CLexerBase.extractPreprocessorOptions(args)

        sourceName = getattr(input, 'name', None) or getattr(input, 'fileName', None) or ""
        inputText = input.getText(0, input.size - 1)

        if not sourceName or not sourceName.endswith(".c"):
            sourceName = "stdin.c"

        outputName = sourceName + ".p"

        if nopp:
            try:
                with open(outputName, "w") as f:
                    f.write(inputText)
            except:
                pass
            return InputStream(inputText)

        if sourceName == "stdin.c":
            with open(sourceName, "w") as f:
                f.write(inputText)

        if gcc:
            output = ""
            try:
                gccCommand = "gcc.exe" if is_windows else "gcc"
                ppOptsStr = " ".join('"' + o + '"' for o in ppOptions)
                cmd = f'{gccCommand} -std=c2x -E -C {ppOptsStr} "{sourceName}"'
                result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
                output = result.stdout
            except:
                pass
            with open(outputName, "w") as f:
                f.write(output)
            return InputStream(output)

        if clang:
            output = ""
            try:
                clangCommand = "clang.exe" if is_windows else "clang"
                ppOptsStr = " ".join('"' + o + '"' for o in ppOptions)
                cmd = f'{clangCommand} -std=c2x -E -C {ppOptsStr} "{sourceName}"'
                result = subprocess.run(cmd, shell=True, capture_output=True, text=True)
                output = result.stdout
            except:
                pass
            with open(outputName, "w") as f:
                f.write(output)
            return InputStream(output)

        raise Exception("No preprocessor specified.")

    @staticmethod
    def extractPreprocessorOptions(args):
        options = []
        for arg in args:
            if arg.startswith("--D"):
                options.append("-D" + arg[3:])
            elif arg.startswith("--I"):
                options.append("-I" + arg[3:])
        return options
